File: DPC++/sycl-benchmark/fibonacci.cpp

module;

/// <summary>
/// Fibonacci module contains differnt implemntation to calculate Fibbonacci numbers
/// </summary>
module fibonacci;

namespace fib {

	/// <summary>
	/// Naive recursive implementation of Fibbonacci numbers
	/// </summary>
	/// <param name="n">Index of Fibonacci number to be calculated</param>
	/// <returns>nth FIbonacci number</returns>
	int fib_naive(const int n) {

		if (n < 2) {
			return n;
		}
		else {
			return fib_naive(n - 1) + fib_naive(n - 2);
		}
		return 0;
	}

	/// <summary>
	/// Helper for tail-call optimized version
	/// </summary>
	int fib_aux(const int a, const int b, const int n) {

		if (n < 1) {
			return a;
		}
		else {
			return fib_aux(b, a + b, n - 1);
		}
	}

	/// <summary>
	/// Tail-call optimized implementation of Fibonacci numbers
	/// </summary>
	/// <param name="n">Index of Fibonacci number to be calculated</param>
	/// <returns>nth FIbonacci number</returns>
	int fib(const int n)
	{
		return fib_aux(0, 1, n);
	}
}

File: Csharp/Benchmark/Benchmark.Tasks/Fibonacci.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Numerics;
using System.Text;
using System.Threading.Tasks;

namespace Benchmark.Tasks
{
    /// <summary>
    ///  A collection of various implementations to calculate Fibonacci numbers
    /// </summary>
    public class Fibonacci
    {
        /// <summary>
        /// Naive recursive implementation
        /// </summary>
        /// <param name="n">nth Fibonacci number to be calculated</param>
        /// <returns>Calculated nth Fibonacci number</returns>
        public static int NaiveFib(int n)
        {
            if (n < 2) return n;
            else
                return NaiveFib(n - 1) + NaiveFib(n - 2);
        }

        /// <summary>
        /// Internal recursive loop fro FIbonacci number calculation
        /// </summary>
        /// <typeparam name="T">Numeric result type for calculated Fibonacci number</typeparam>
        /// <param name="n">nth Fibonacci number to be calculated</param>
        /// <param name="a">First intermediate sum variable</param>
        /// <param name="b">Second intermediate sum variable</param>
        /// <returns></returns>
        private static T LoopFib<T>(int n, T a, T b)
            where T : System.Numerics.IAdditionOperators<T, T, T>
        {
            if (n == 0)
            {
                return a;
            }
            else
            {
                return LoopFib(n - 1, b, a + b);
            }

        }

        /// <summary>
        /// Tail-call optimized recursive implementation
        /// </summary>
        /// <param name="n">nth Fibonacci number to be calculated</param>
        /// <returns>Calculated nth Fibonacci number</returns>
        public static int RecFib(int n)
        {
            return LoopFib(n, 0, 1);
        }

        /// <summary>
        /// Tail-call optimized recursive implementation, using BigInteger for result
        /// </summary>
        /// <param name="n">nth Fibonacci number to be calculated</param>
        /// <returns>Calculated nth Fibonacci number</returns>
        public static BigInteger RecFibBigInt(int n)
        {
            return LoopFib(n, BigInteger.Zero, BigInteger.One);
        }

    }

}
